File: python/app/pubsub.py
from __future__ import annotations

import queue
import threading
from typing import Callable

from .errors import PubSubClosedError
from .models import Event


class Subscription:
    def __init__(self, buffer_size: int) -> None:
        self.queue: queue.Queue[Event] = queue.Queue(maxsize=buffer_size)
        self.closed = threading.Event()


class PubSub:
    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._closed = False
        self._subscription: Subscription | None = None

    def publish(self, event: Event) -> None:
        with self._lock:
            if self._closed:
                raise PubSubClosedError("pubsub is closed")
            subscription = self._subscription

        if subscription is None or subscription.closed.is_set():
            return

        while True:
            if subscription.closed.is_set():
                return
            try:
                subscription.queue.put(event, timeout=0.1)
                return
            except queue.Full:
                continue

    def subscribe(self, buffer_size: int) -> tuple[Subscription, Callable[[], None]]:
        if buffer_size <= 0:
            buffer_size = 1
        subscription = Subscription(buffer_size)

        with self._lock:
            if self._closed or self._subscription is not None:
                subscription.closed.set()
                return subscription, lambda: None
            self._subscription = subscription

        def cancel() -> None:
            with self._lock:
                if self._subscription is subscription:
                    self._subscription = None
                    subscription.closed.set()

        return subscription, cancel

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._closed = True
            if self._subscription is not None:
                self._subscription.closed.set()
                self._subscription = None

File: node/src/domain/pubsub.ts
import { PubSubClosedError } from "./errors";
import { Event } from "./models";

class AsyncQueue<T> {
  private items: T[] = [];
  private waiters: Array<(result: IteratorResult<T>) => void> = [];
  private spaceWaiters: Array<() => void> = [];
  private closed = false;

  constructor(private maxSize: number) {}

  async push(item: T): Promise<void> {
    if (this.closed) {
      throw new PubSubClosedError("subscription is closed");
    }

    while (this.items.length >= this.maxSize && !this.closed) {
      await new Promise<void>((resolve) => this.spaceWaiters.push(resolve));
    }

    if (this.closed) {
      throw new PubSubClosedError("subscription is closed");
    }

    const waiter = this.waiters.shift();
    if (waiter) {
      waiter({ value: item, done: false });
      return;
    }

    this.items.push(item);
  }

  async shift(): Promise<IteratorResult<T>> {
    if (this.items.length > 0) {
      const item = this.items.shift() as T;
      this.notifySpace();
      return { value: item, done: false };
    }

    if (this.closed) {
      return { value: undefined as T, done: true };
    }

    return new Promise<IteratorResult<T>>((resolve) => {
      this.waiters.push(resolve);
    });
  }

  close(): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    while (this.waiters.length > 0) {
      const resolve = this.waiters.shift();
      if (resolve) {
        resolve({ value: undefined as T, done: true });
      }
    }
    this.notifyAllSpace();
  }

  private notifySpace(): void {
    const resolve = this.spaceWaiters.shift();
    if (resolve) {
      resolve();
    }
  }

  private notifyAllSpace(): void {
    while (this.spaceWaiters.length > 0) {
      const resolve = this.spaceWaiters.shift();
      if (resolve) {
        resolve();
      }
    }
  }
}

export class Subscription implements AsyncIterable<Event> {
  constructor(private queue: AsyncQueue<Event>) {}

  [Symbol.asyncIterator](): AsyncIterator<Event> {
    return {
      next: () => this.queue.shift(),
    };
  }
}

export class PubSub {
  private closed = false;
  private queue: AsyncQueue<Event> | null = null;

  async publish(event: Event): Promise<void> {
    if (this.closed) {
      throw new PubSubClosedError("pubsub is closed");
    }
    if (!this.queue) {
      return;
    }
    await this.queue.push(event);
  }

  subscribe(bufferSize: number): { subscription: Subscription; cancel: () => void } {
    const size = bufferSize > 0 ? bufferSize : 1;
    const queue = new AsyncQueue<Event>(size);

    if (this.closed || this.queue) {
      queue.close();
      return { subscription: new Subscription(queue), cancel: () => {} };
    }

    this.queue = queue;

    const cancel = () => {
      if (this.queue === queue) {
        this.queue = null;
      }
      queue.close();
    };

    return { subscription: new Subscription(queue), cancel };
  }

  close(): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    if (this.queue) {
      this.queue.close();
      this.queue = null;
    }
  }
}
